File: frontend/src/types/workflow.ts
export type WorkflowStatus = 'draft' | 'active' | 'inactive' | 'archived';
export type StepType = 'agent' | 'tool' | 'condition' | 'loop' | 'parallel' | 'wait';

export interface WorkflowStepConfig {
  agentId?: string;
  toolId?: string;
  condition?: string;
  loopCount?: number;
  loopCondition?: string;
  waitDuration?: number;
  parallelSteps?: string[];
  inputMapping?: Record<string, string>;
  outputMapping?: Record<string, string>;
}

export interface WorkflowStep {
  id: string;
  name: string;
  type: StepType;
  config: WorkflowStepConfig;
  position: { x: number; y: number };
  nextSteps: string[];
  onError?: 'stop' | 'continue' | 'retry';
  retryCount?: number;
}

export interface Workflow {
  id: string;
  name: string;
  description: string;
  status: WorkflowStatus;
  steps: WorkflowStep[];
  variables: Record<string, unknown>;
  triggers?: WorkflowTrigger[];
  createdAt: string;
  updatedAt: string;
  executionCount: number;
  lastExecutedAt?: string;
}

export interface WorkflowTrigger {
  id: string;
  type: 'schedule' | 'webhook' | 'event' | 'manual';
  config: Record<string, unknown>;
  enabled: boolean;
}

export interface CreateWorkflowRequest {
  name: string;
  description: string;
  steps?: WorkflowStep[];
  variables?: Record<string, unknown>;
  triggers?: WorkflowTrigger[];
}

export interface UpdateWorkflowRequest {
  name?: string;
  description?: string;
  status?: WorkflowStatus;
  steps?: WorkflowStep[];
  variables?: Record<string, unknown>;
  triggers?: WorkflowTrigger[];
}

File: backend/app/models/workflow.py
"""
Workflow model.
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class WorkflowStep(BaseModel):
    """A step in a workflow."""

    id: str
    name: str
    type: str  # "agent", "tool", "condition", "parallel"
    config: Dict[str, Any] = {}
    next_steps: List[str] = []


class Workflow(BaseModel):
    """Workflow model representing a sequence of agent/tool operations."""

    id: Optional[str] = Field(None, alias="_id")
    name: str
    description: str
    status: str = "draft"  # "draft", "active", "archived"
    steps: List[WorkflowStep] = []
    trigger: Dict[str, Any] = {}  # "manual", "scheduled", "event"
    config: Dict[str, Any] = {}
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    class Config:
        populate_by_name = True
